cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -91,6 +103,126 @@ func stampedAction(stamp string, action func() error) error {
 	return ioutil.WriteFile(stampFile, nil, 0644)
 }
 
+// SystemdMountOptions reflects the set of options for mounting something using
+// systemd-mount(1)
+type SystemdMountOptions struct {
+	// is tmpfs indicates that "what" should be ignored and a new tmpfs should
+	// be mounted at the location
+	IsTmpfs bool
+	// SurvivesPivotRoot indicates that the mount should persist from the
+	// initramfs to after the pivot_root to normal userspace
+	// this is done by creating systemd unit .wants symlinks on
+	// initrd-device.target in /run
+	// TODO: all current instances of doSystemdMount set this to true, should we
+	// invert or just drop the option and make it the default behavior ?
+	SurvivesPivotRoot bool
+	// NeedsFsck indicates that before returning to the caller, an fsck check
+	// should be performed on the thing being mounted
+	NeedsFsck bool
+	// NoWait will not wait until the systemd unit is active and running, which
+	// is the default behavior
+	NoWait bool
+}
+
+var (
+	doSystemdMount = doSystemdMountImpl
+	counter        = 1
+)
+
+// doSystemdMount will mount "what" at "where" using systemd-mount(1) with
+// various options
+func doSystemdMountImpl(what, where string, opts *SystemdMountOptions) error {
+	if opts == nil {
+		opts = &SystemdMountOptions{}
+	}
+
+	// doesn't make sense to fsck a tmpfs
+	if opts.NeedsFsck && opts.IsTmpfs {
+		return fmt.Errorf("cannot mount %q at %q: impossible to fsck a tmpfs", what, where)
+	}
+
+	whereEscaped := systemd.EscapeUnitNamePath(where)
+	unitName := whereEscaped + ".mount"
+
+	args := []string{"--no-pager", "--no-ask-password", what, where}
+	if opts.IsTmpfs {
+		args = append(args, "--type=tmpfs")
+	}
+
+	if opts.NeedsFsck {
+		args = append(args, "--fsck=yes")
+	}
+
+	// TODO: do we really need this? unclear why we would need to ever not block
```

Comment:
If there are no systemd transaction in progress one can block. Otherwise, the block will be indefinite, if the unit that is starting enqueus a job, which will not be started until after the call completes and the queue is unblocked.

This was needed before, when the-tool.service was oneshot instance unit, that was supposed to exit, for systemd-mount jobs to be started.

Reply:
I think the desired behavior now is that "the-tool.service" runs as late as possible in the initrd, namely after systemd-modules-load and systemd-udev, etc. all have run, such that nothing (except perhaps populate-writable.service) depends on `snap-bootstrap initramfs-mounts`. With that in place, then I don't think there is any concern about any mount units we create from `snap-bootstrap initramfs-mounts` depending on something which will not be started, so I think in that world we don't ever need to not block.

Thanks for clarifying, I will remove this TODO

Reply:
I think it's a good idea to add a comment with Dimitri's explanation here.